src/Npgsql/NpgsqlConnection.cs

```diff
@@ -38,6 +38,11 @@ public sealed class NpgsqlConnection : DbConnection, ICloneable, IComponent
     // Set this when disposed is called.
     bool _disposed;
 
+    // Set to true when the private NpgsqlDataSource is created by setting the ConnectionString property (legacy code path)
+    // in which case we want to retry if the DataSource has been disposed by the PoolManager as a reaction
+    // to Clean() or CleanAll()
+    bool _retryDisposedDataSource;
```

Comment:
Maybe it's better to move this property to the `DataSource`?

Reply:
> Maybe it's better to move this property to the DataSource?

Really? Why?
Looking at the current code it would make everything slightly more complicated.

Reply:
Since the thing we want to determine is whether the data source we reference is created from settings (legacy) or builder, to me it makes sense to have that flag on the DataSource and set it accordingly (to `false` if the DataSource is created from builder, `true` otherwise). And if that property is internal, that shouldn't make things hard (you're just going to check it instead of a private field).

Reply:
And the issue with the private field we already have is that it's mutable, which can lead to problems if we're able to first reference a legacy DataSource, but after that somehow reference a non-legacy one.

Reply:
> Since the thing we want to determine is whether the data source we reference is created from settings (legacy) or builder, to me it makes sense to have that flag on the DataSource and set it accordingly (to false if the DataSource is created from builder, true otherwise).

In the end, all `NpgsqlDataSource` instances are created via `DataSourceBuilder`. For the legacy ones it happens in [`SetupDataSource()`](https://github.com/npgsql/npgsql/blob/c4481066fa757aa3de27020407d851555b6f3b1e/src/Npgsql/NpgsqlConnection.cs#L222).

We're currently lacking a way to inform a DataSource about it's legacy status and a settable property wouldn't address your mutablity concerns, so this flag has to make its way from `SetupDataSource()` to the `DataSourceBuilder` and then to `NpgsqlDataSource`. That's essentially what I meant by "slightly more complicated" but I can definitely do this.